### ISSUE TITLE
[API Proposal]: IKeyedServiceProvider Have a method to get all keys for a ServiceType, an option to make the keys services unique and the option to getservices return all keyed services

### ISSUE THREAD

**Issue by luizfbicalho** [dotnet/runtime#91466]
### Background and motivation

I have some configurations in my process that I create for each step one service and store it in a EF class

```C#

    public class Step
    {
        public int Id { get; set; }
        public string StepName { get; set; }
        public string StepKeyService { get; set; }

        public IService1 GetService(IServiceProvider provider) {

            return provider.GetRequiredKeyedService<IService>(StepKeyService);
        }
    }
```

But to make the Editor for the steps I need to list all of the Keys that are registered in the ServiceProvider

One second utility would be to Make the Service Type and Key marked as unique
```c#
col.SetKeyedServiceUnique<IService>()

```

this way if I add two services I would get a validation exeption

```C#
            col.AddKeyedTransient<IService, ServiceA>(KeyedService.AnyKey);
            col.AddKeyedTransient<IService, ServiceB>(KeyedService.AnyKey);
```

and the third question is to have a method provider.GetAllServices<T>() that return all services from all keys
that could be a keyed dictionary if it would help organize

### API Proposal

```csharp


public static class  ServiceProviderExtensions
{
     // prevent this keyed service to have more than one implementation
     public void SetKeyedServiceUnique<T>(this IServiceCollection collection);
     // get all services from all keys
     public IEnumerable<T> GetAllServices<T>(this IKeyedServiceProvider provider);
     // get all services and all keys
     public IDictionary<object?,IEnumerable<T>> GetAllServicesDictionary<T>(this IKeyedServiceProvider provider);
     // get all keys
     public IEnumerable<object?> GetAllKeys<T>(this IKeyedServiceProvider provider);
     
}
```

### API Usage

```csharp
     [ApiController]
    public class StepController : ControllerBase

    {
        public StepController(IServiceProvider provider)
        {
            Provider = provider;
        }

        public IServiceProvider Provider { get; }

        public ActionResult<IEnumerable<object?>> GetStepKeys()
        {
            return Ok(Provider.GetAllKeys<IService>();
        }
    }   
```

### Alternative Designs

I'm open to any alternative that could result in this functionalities

### Risks

I can't imagine risks because there are new methods to implement this.

**Comment by Cassinotte** [comment]
I would also like an alternative to the question

**Comment by steveharter** [comment]
@benjaminpetit any thoughts on the keyed-service ask for make the Service Type and Key unique along with validation?

**Comment by steveharter** [comment]
> But to make the Editor for the steps I need to list all of the Keys that are registered in the ServiceProvider

For this scenario, is it possible to enumerate the ServiceCollection instead which can return the ServiceDescriptors?

Also since there can be several implementations of IServiceProvider\IKeyedServiceProvider, that has to be accounted for in any extension methods -- i.e. they need to be implemented in terms of types exposed in the `Microsoft.Extensions.DependencyInjection.Abstractions` assembly, not the types in `Microsoft.Extensions.DependencyInjection` for example. This likely means we'd have to add new interfaces as well to expose the keys.

**Comment by luizfbicalho** [comment]
> > But to make the Editor for the steps I need to list all of the Keys that are registered in the ServiceProvider
> 
> For this scenario, is it possible to enumerate the ServiceCollection instead which can return the ServiceDescriptors?
> 
> Also since there can be several implementations of IServiceProvider\IKeyedServiceProvider, that has to be accounted for in any extension methods -- i.e. they need to be implemented in terms of types exposed in the `Microsoft.Extensions.DependencyInjection.Abstractions` assembly, not the types in `Microsoft.Extensions.DependencyInjection` for example. This likely means we'd have to add new interfaces as well to expose the keys.

How can i get the service colector from the service provider?

**Comment by benjaminpetit** [comment]
> @benjaminpetit any thoughts on the keyed-service ask for make the Service Type and Key unique along with validation?

If we add this, then we should add the same method for non-keyed DI in my opinion.

I think for this and the listening of all registered key, you could implement your own `IServiceProviderFactory` that will validate that you don't have two or more services registered with the same key, and build a list of available keys. You can then insert this list as a new service to the `IServiceCollection` before passing it to the "real" IServiceProviderFactory` implementation.

Does that make sense?

**Comment by luizfbicalho** [comment]
> If we add this, then we should add the same method for non-keyed DI in my opinion.

Would it be useful to add something like this

`serviceCollection.AddScoped<IBar, Bar>().AsUnique();`


Or add some more validation with a func

`serviceCollection.AddScoped<IBar, Bar>().WithValidation(collection=>VerifySomething(collection));
`

**Comment by benjaminpetit** [comment]
`ServiceProviderOptions.ValidateOnBuild` should already allow you to check that your services are instantiable.

Otherwise, I think everything you want to do is doable with some kind of decorator around `IServiceProviderFactory`

**Comment by luizfbicalho** [comment]
Can you provide any example like this? How to validate with the factory

I'm interested in the @steveharter idea of get the service collection from the service provider

**Comment by benjaminpetit** [comment]
Here is an example on how to generate a dictionary of all keyed services: https://gist.github.com/benjaminpetit/468741882f1ad6e4ec8dad761103a87d

**Comment by luizfbicalho** [comment]
Thanks a lot @benjaminpetit 

Ano ideal on how to get the ServiceCollection from the ServiceProvider scoped instance?

**Comment by luizfbicalho** [comment]
Thanks a lot @benjaminpetit , I'll try to transform this in a usefull library

**Comment by stephentoub** [comment]
I just hit this hard as well. I needed to be able to get all of the instances for a particular service type (I didn't actually need the keys), and I thought `GetKeyedServices(KeyedService.AnyKey)` would work, but it doesn't, as the key matching doesn't factor AnyKey in for the key being searched.  This is an unfortunate gap.

**Comment by thomaslevesque** [comment]
I was going to use keyed services for the first time, and also hit this limitation. I can work around it, but this gap should be closed IMO.

**Comment by davhdavh** [comment]
IMHO, this is more of a bug. Documentation for `KeyedService.AnyKey` specifically says:
```
Represents a key that matches any key.
```
So `GetKeyedServices<T>(KeyedService.AnyKey)` should most definitely return an enumeration of all `T` implementations.
Similar, `col.AddKeyedTransient<IService, ServiceA>(KeyedService.AnyKey);` should give an error. If you want to have a fallback, `null` should be used.

**Comment by robertmclaws** [comment]
Hey folks! Just wanted to find out is this was planning on getting fixed at all? There shouldn't need to be hacks to make this work, since `KeyedService.AnyKey` exists and should be implemented. Thanks!

**Comment by 0xced** [comment]
There's a discussion going on at #100105 (which is a duplicate of this issue).

**Comment by michael-hawker** [comment]
Yeah, I have a scenario where I need to register services to get pulled specifically for different use cases to load specific types of data, but also need to gather them all (thus the common interface) to initialize them beforehand.

I got excited by the new keyed services as it seemed built for this scenario, but then got stuck in the last step after changing my code in trying to get all the services. I was hoping I could just call `GetKeyServices<T>()` to get them all, I didn't even find the docs on `KeyedService.AnyKey`, but then that seemed great too (as called out above)... but then that also didn't work. So, I think I'm back to being stuck without hacky workarounds...

**Comment by julealgon** [comment]
@michael-hawker 
> Edit: Oh, just saw the `KeyedService.AnyKey` was fixed in .NET 9 [#109016](https://github.com/dotnet/runtime/issues/109016) 🎉🎉🎉, hopefully it'll get backported too. (Just a note for anyone here, you have to ensure to update your NuGet package to the latest .NET 9 version too, **not just your TFM**.)

I don't think you even _need_ to update your TFM. Isn't this a behavior of the library itself, `Microsoft.Extensions.DependencyInjection`? Upgrading the library to v9 will be enough even if you are still using .NET 8, since the fix doesn't rely on any `#if` checks against a specific TFM.

**Comment by michael-hawker** [comment]
@julealgon indeed, thanks! It was late at night, so I had bumped the TFM first before realizing it was just a package dependency. 😅 Worked like a charm!

**Comment by mrek-msft** [comment]
> But I think that teh option to validate or enumerate che registrations before the build could help out

Something like https://github.com/dotnet/runtime/issues/119646 proposal?
